File: packages/cordova-consent/src/android/ExecuteContext.java
package cordova.plugin.consent;

import android.app.Activity;

import com.google.android.ump.ConsentDebugSettings;
import com.google.android.ump.ConsentRequestParameters;

import org.apache.cordova.CallbackContext;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.Objects;

public class ExecuteContext {
    public static Consent plugin;
    public final String actionKey;
    public final JSONArray args;
    public final CallbackContext callbackContext;
    public final JSONObject opts;

    ExecuteContext(String actionKey, JSONArray args, CallbackContext callbackContext) {
        this.actionKey = actionKey;
        this.args = args;
        this.callbackContext = callbackContext;
        this.opts = args.optJSONObject(0);
    }

    public int optId() {
        return opts.optInt("id");
    }

    public ConsentRequestParameters optConsentRequestParameters() {
        ConsentRequestParameters.Builder builder = new ConsentRequestParameters.Builder();
        if (this.opts == null) {
            return builder.build();
        }

        if (this.opts.has("tagForUnderAgeOfConsent")) {
            builder.setTagForUnderAgeOfConsent(this.opts.optBoolean("tagForUnderAgeOfConsent"));
        }

        builder.setConsentDebugSettings(optConsentDebugSettings());

        return builder.build();
    }

    public ConsentDebugSettings optConsentDebugSettings() {
        ConsentDebugSettings.Builder builder = new ConsentDebugSettings.Builder(getActivity());

        if (opts.has("debugGeography")) {
            builder.setDebugGeography(opts.optInt("debugGeography"));
        }

        if (opts.has("testDeviceIds")) {
            JSONArray ids = opts.optJSONArray("testDeviceIds");
            for (int i = 0; i < Objects.requireNonNull(ids).length(); i++) {
                String testDeviceId = ids.optString(i);
                if (testDeviceId != null) {
                    builder.addTestDeviceHashedId(testDeviceId);
                }
            }
        }

        return builder.build();
    }

    private Activity getActivity() {
        return plugin.cordova.getActivity();
    }
}

File: packages/cordova-consent/src/android/ExecuteContext.kt
package cordova.plugin.consent

import android.app.Activity
import com.google.android.ump.ConsentDebugSettings
import com.google.android.ump.ConsentInformation
import com.google.android.ump.ConsentRequestParameters
import org.apache.cordova.CallbackContext
import org.apache.cordova.PluginResult
import org.json.JSONArray
import org.json.JSONObject
import java.util.Objects

data class ExecuteContext(
    val actionKey: String,
    val args: JSONArray,
    val callbackContext: CallbackContext,
    val plugin: Consent,
) {
    val opts: JSONObject? = args.optJSONObject(0)

    fun optId(): Int {
        return opts!!.optInt("id")
    }

    fun optConsentRequestParameters(): ConsentRequestParameters {
        val builder = ConsentRequestParameters.Builder()
        if (opts == null) {
            return builder.build()
        }
        if (opts.has("tagForUnderAgeOfConsent")) {
            builder.setTagForUnderAgeOfConsent(opts.optBoolean("tagForUnderAgeOfConsent"))
        }
        builder.setConsentDebugSettings(optConsentDebugSettings())
        return builder.build()
    }

    fun resolve() {
        callbackContext.success()
    }

    fun resolve(data: Boolean) {
        val result = PluginResult(PluginResult.Status.OK, data)
        sendResult(result)
    }

    fun resolve(status: ConsentInformation.PrivacyOptionsRequirementStatus) {
        callbackContext.success(status.ordinal)
    }

    private fun sendResult(result: PluginResult?) {
        callbackContext.sendPluginResult(result)
    }

    private fun optConsentDebugSettings(): ConsentDebugSettings {
        val builder = ConsentDebugSettings.Builder(activity)
        if (opts!!.has("debugGeography")) {
            builder.setDebugGeography(opts.optInt("debugGeography"))
        }
        if (opts.has("testDeviceIds")) {
            val ids = opts.optJSONArray("testDeviceIds")
            for (i in 0 until Objects.requireNonNull(ids).length()) {
                val testDeviceId = ids?.optString(i)
                if (testDeviceId != null) {
                    builder.addTestDeviceHashedId(testDeviceId)
                }
            }
        }
        return builder.build()
    }

    private val activity: Activity
        get() = plugin.cordova.activity
}
